onnxruntime/core/graph/function_utils.cc

```diff
@@ -0,0 +1,672 @@
+  }
+
+  auto* onnx_func_proto = iter->second;
+  ORT_ENFORCE(onnx_func_proto);
+  // generate the schema for this function template
+  // For schema defined functions get the version from the node in parent graph.
+  // For the functions which do not have schema defined (model local functions)
+  // get the since version from the version in opset imports using the domain.
+  auto since_version = GetVersionForDomain(function_domain, domain_version_map);
+  auto op_schema = std::make_unique<ONNX_NAMESPACE::OpSchema>();
+  op_schema->SetName(function_name);
+  op_schema->SetDomain(function_domain);
+  op_schema->SetDoc(onnx_func_proto->doc_string());
+  op_schema->SinceVersion(static_cast<ONNX_NAMESPACE::OperatorSetVersion>(since_version));
+  std::unordered_map<std::string, int> input_name_idx_map;
+  std::unordered_map<std::string, int> output_name_idx_map;
+
+  for (int i = 0; i < onnx_func_proto->input_size(); ++i) {
+    input_name_idx_map[onnx_func_proto->input().Get(i)] = i;
+  }
+  for (int i = 0; i < onnx_func_proto->output_size(); ++i) {
+    output_name_idx_map[onnx_func_proto->output().Get(i)] = i;
+  }
+
+  // Infer a op_schema for stand-alone functions.
+  IOTypeConstraintHelper(*onnx_func_proto, op_schema, input_name_idx_map, output_name_idx_map);
+  auto allow_official_onnx_release_only_final =
+      allow_released_opsets_only && model_load_utils::IsAllowReleasedONNXOpsetsOnlySet();
+
+  const auto onnx_released_versions =
+      schema_registry.GetLastReleasedOpsetVersions(false);
+
+  std::unordered_map<std::string, int> func_domain_to_version;
+  for (auto& opSet : onnx_func_proto->opset_import()) {
+    const auto& domain = opSet.domain();
+    const auto version = gsl::narrow_cast<int>(opSet.version());
+
+    model_load_utils::ValidateOpsetForDomain(onnx_released_versions, logger,
+                                             allow_official_onnx_release_only_final, domain, version);
+
+    // We need to overwrite the domain here with ("") or else the loop below will try to find ("")
+    // in the map and if not found (when domain == kOnnxDomainAlias), adds an entry for ("", 11).
+    // This effectively ignores the opset version specified by the model for the onnx domain.
+    if (domain == kOnnxDomainAlias) {
+      func_domain_to_version[kOnnxDomain] = version;
+    } else {
+      func_domain_to_version[domain] = version;
+    }
+  }
+
+  op_schema->TypeAndShapeInferenceFunction(
+      [onnx_func_proto, func_domain_to_version, &model_local_functions](ONNX_NAMESPACE::InferenceContext& ctx) {
+        auto schema_registry = ONNX_NAMESPACE::OpSchemaRegistry::Instance();
+        ONNX_NAMESPACE::ShapeInferenceOptions options{true, 1, false};
+        InferShapeForFunctionNode(ctx, *onnx_func_proto, func_domain_to_version, options, schema_registry, model_local_functions, function_utils::GetFunctionIdentifier);
```

Comment:
When if the "model_local_functions" object is already gone when this lambda function is running? How do you ensure it won't happen?

Reply:
that is another piece i don't like it... i just move the existing code here, haven't got time to refactor this part.
The existing design is, the "model_local_functions" map is hold by graph, and graph is hold by model. This generated custom op schema is also hold by model. Since both graph and generated schema are hold by model, they will have the same life cycle, so it is safe. BUT, if someone accidently reset the graph in the model, it will break.

I actually don't like the idea that generate a temporary schema for those function nodes, it introduces un-necessary complexity. but we have too many codes make assumption that every node should have a schema, change it need quite big efforts...

Reply:
update the implementation to host the model local function map together with generated op schema in model